core/src/main/java/org/apache/iceberg/rest/PaginatedList.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.function.Supplier;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.rest.responses.ListNamespacesResponse;
+import org.apache.iceberg.rest.responses.ListTablesResponse;
+import org.apache.iceberg.rest.responses.Route;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PaginatedList<T> implements List<T> {
```

Comment:
I have some concerns about the way we're implementing the pagination.  It doesn't appear like we're properly adhering to the List interface behavior.  For example, if we perform a `namespaces = catalog.listNamespaces(...)` and then check `namespaces.size()` we would get zero even if there are results.  Other methods like `contains` would appear to not have the right behavior either.

Overall, it seems like we're trying to hide the pagination behind the List interface but not really getting the behavior we want.  We might want to consider just prefetching as it seems the primary limitation is expected to be on the server side (client will likely need to fit the whole result in memory anyway).  That may simplify this and not require the paginated list implementation.

Reply:
@danielcweeks Can you elaborate on what you mean by prefetching? Did you want to the client to retrieve all the paginated results from the service in one go ( by repeatedly sending requests until `next-page-token` is `null`)?

I had this before:

```
 List<Namespace> namespaces = Lists.newArrayList();
      String pageToken = "";
      queryParams.put("pageSize", restPageSize);
 
      while (pageToken != null) {
        queryParams.put("pageToken", pageToken);
        LOG.error("Sending list namespaces request");
        ListNamespacesResponse response =
                client.get(
                        paths.namespaces(),
                        queryParams,
                        ListNamespacesResponse.class,
                        headers(context),
                        ErrorHandlers.namespaceErrorHandler());
        pageToken = response.nextPageToken();
        namespaces.addAll(response.namespaces());
      }
```

 but i think the other reviewers of the pr mentioned that we want to do lazily loading of the results, and shared this `PaginatedList` as an idea https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/dynamodbv2/datamodeling/PaginatedList.html which is what i tried following.

Reply:
To clarify my comment, I think there are really two issues:

1) The current implementation does not properly adhere to the List behaviors.  The dynamodb example that's referenced will eagerly fetch if certain methods are called, which is necessary to get the correct behavior out of the List interface.  

2) The second issue is more about the utility of a lazy pagination since it seems the original motivation is more to address server side limitations (like Glue's 100 results per request limit) as opposed to a client side memory limitation.  While you might be able to optimize something like the `SHOW TABLES LIKE` command client-side, we haven't see the client-side as the limiting factor.

Reply:
+1, Current implementation of paginated list seems wrong as caller can get unexpected results. Two possible options are:
1. Pre-fetch all items (as suggested above)
2. methods like size(), contains(), isEmpty(), toArray(), get() .. first paginates across all items before performing the operation.

Reply:
@danielcweeks @sachet Have done a revision for this using a prefetch approach if you can take a look whenever would be ideal.